cmd/dp-recipe-api/main.go

```diff
@@ -64,4 +67,36 @@ func main() {
 		api.CreateAndInitialiseRecipeAPI(ctx, *cfg, store.DataStore{Backend: nil})
 	}
 
+	apiErrors := make(chan error, 1)
+
+	// Gracefully shutdown the application closing any open resources.
+	gracefulShutdown := func() {
+		log.Event(ctx, (fmt.Sprintf("shutdown with timeout: %s", cfg.GracefulShutdownTimeout)), log.INFO)
+		ctx, cancel := context.WithTimeout(context.Background(), cfg.GracefulShutdownTimeout)
+
+		// stop any incoming requests before closing any outbound connections
+		api.Close(ctx)
+
+		if enableMongoData {
+
+			if err := mongolib.Close(ctx, mongodb.Session); err != nil {
+				log.Event(ctx, "failed to close mongo session", log.ERROR, log.Error(err))
+			}
+		}
+
+		log.Event(ctx, "shutdown complete", log.INFO)
+
+		cancel()
+		os.Exit(1)
+	}
+
+	for {
+		select {
+		case err := <-apiErrors:
+			log.Event(ctx, "api error received", log.ERROR, log.Error(err))
```

Comment:
should we also do a graceful shutdown if an API error occurs?

Reply:
When I was doing the graceful Shutdown with Nathan, we also look for apiErrors and then ensure that all services are closed gracefully even if the apiError occurred for a specific reason. This is the same in the DatasetAPI as well.